### PR TITLE
[7x]: Added Support for -i flag with differential recovery

### PR DESCRIPTION
**Pull request by kpraveen457**
This PR added support for using the -i flag with differential recovery in the gprecoverseg command, enhancing its capabilities.

Implementation: 
**Simultaneous Use of -i and --differential:**
Added the functionality to allow the use of both the -i and --differential options simultaneously in the gprecoverseg command. This means that by providing a recovery configuration file with the -i flag and specifying the --differential flag, you will initiate a differential recovery process for all hosts listed in the recovery configuration file, similar to how the -i and -F options currently work together

**Recovery Type Specification in Recovery Configuration File:** 
Added the functionality to specify I/D/F as a recovery type for each line in recovery_config_file and based on that recovery type respective recovery will be performed. The aim is to support existing formats of recovery config files, maintain backward compatibility, and add support for differential recovery.
 E.g; A sample config file will look as follows:
```
I|sdw1|50001|/data1/primary/gpseg4  # Does incremental recovery
D|sdw1|50001|/data1/primary/gpseg4  # Does differential recovery
F|sdw1|50001|/data1/primary/gpseg4  # Does full recovery
sdw1|50001|/data1/primary/gpseg4     # Does incremental recovery 
sdw1-1|sdw1|50001|/data1/primary/gpseg4  # Does incremental recovery 
D|sdw1-1|sdw1|50001|/data1/primary/gpseg4  # Does differential recovery 
```

**Test:**
Added unit and behave test cases

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gppylib/programs/clsRecoverSegment.py [hunk 1]
```diff
@@ -234,10 +236,18 @@ def validateRecoveryParams(self):
             optionCnt += 1
         if self.__options.rebalanceSegments:
             optionCnt += 1
+        if optionCnt > 1:
+            raise ProgramArgumentValidationException("Only one of -i, -p and -r and may be specified")
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:240]: [typo]: -> Only one of -i, -p and -r may be specified

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature [hunk 2]
```diff
@@ -2376,3 +2373,44 @@ Feature: gprecoverseg tests
...
+        And all the segments are running
+        And the segments are synchronized
+        And user stops all mirror processes
+        And user can start transactions
+        And the gprecoverseg input file "recover_config_file" is cleaned up
+       When a gprecoverseg input file "recover_config_file" is created with all the failed segments and their recovery type
+        And the user runs "gprecoverseg -i /tmp/recover_config_file -a --differential"
+       Then gprecoverseg should return a return code of 0
+        And gprecoverseg should print "Synchronization mode                 = Differential" to stdout
+        And gprecoverseg should not print "Synchronization mode                 = Full" to stdout
```
Comment (Annu149) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature:2390]: Not a reliable check instead we can modify the above step to check the required string 3 times in log
`And gprecoverseg should print "Synchronization mode                 = Differential" to stdout 3 times`

**Comment by nimish350** [comment]
Tested with the following three scenarios: 

General testing of all three I/F/D values:
```
I|sdw1|7005|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror1/demoDataDir0 
F|sdw1|7006|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror2/demoDataDir1
D|sdw1|7007|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror3/demoDataDir2
```

Default full recovery for one segment
```
sdw1|7005|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror1/demoDataDir0 sdw1|7005|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror1/demoDataDir0 
I|sdw1|7006|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror2/demoDataDir1
D|sdw1|7007|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror3/demoDataDir2
```

Default incremental recovery for one segment:
```
sdw1|7005|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror1/demoDataDir0 
F|sdw1|7006|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror2/demoDataDir1
D|sdw1|7007|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror3/demoDataDir2
```

**Comment by nimish350** [comment]
@Annu149 @ShirishaRao Can we create an proper exception if the value is specified as i/f/d instead of I/F/D

The error it currently provides is ambiguous

```
20231129:13:13:28:017976 gprecoverseg:mdw:gpadmin-[CRITICAL]:-gprecoverseg failed. (Reason='A segment to recover was not found in configuration. This segment is described by 'i|mdw|7005|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror1/demoDataDir0'') exiting...
```

**Comment by Annu149** [comment]
> @Annu149 @ShirishaRao Can we create an proper exception if the value is specified as i/f/d instead of I/F/D
> 
> The error it currently provides is ambiguous
> 
> ```
> 20231129:13:13:28:017976 gprecoverseg:mdw:gpadmin-[CRITICAL]:-gprecoverseg failed. (Reason='A segment to recover was not found in configuration. This segment is described by 'i|mdw|7005|/Users/gpadmin/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast_mirror1/demoDataDir0'') exiting...
> ```
the first element is considered as hostname if it is not in {"I", "D", "F"} hence the error. Hostname can be any string so we can't include any checks also. Should we support i,d,f without matching case?

File: gpMgmt/bin/gppylib/utils.py [hunk 3]
```diff
@@ -640,3 +640,31 @@ def validateHostnameAddress(hostname, address):
...
+    """
+    Extracts relevant recovery configuration information from a list of parts.
+    """
+    address, port, datadir = parts[-3:]
+    recovery_type = "Incremental"
+    hostname_check_required = False
+
+    if len(parts) == 4:
+        first_element = parts[0]
+        if first_element in {"I", "D", "F", "i", "d", "f"}:
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/utils.py:654]: When len(parts) is 4, then part[0] can be either a 'recovery_type' or 'hostname'.
With above approach if hostname is I/D/F/i/d/f (unlikely but still possible), then we are treating hostname as recovery_type. 
Should we not check the validity of hostname first ? If hostname is not valid then we can check if it is in (I,D,F,i,d,f).
With this approach, we will cover all the possibilities.

File: gpMgmt/bin/gppylib/programs/clsRecoverSegment.py [hunk 4]
```diff
@@ -278,14 +280,13 @@ def validateRecoveryParams(self):
...
             optionCnt += 1
-        if self.__options.recoveryConfigFile is not None:
-            optionCnt += 1
         if self.__options.rebalanceSegments:
             optionCnt += 1
-        if self.__options.differentialResynchronization:
-            optionCnt += 1
-        if optionCnt > 1:
-            raise ProgramArgumentValidationException("Only one of -i, -p, -r and --differential may be specified")
+        if optionCnt > 0 and self.__options.recoveryConfigFile is not None:
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:285]: Validation of option '-p' with '-r' is missing and needs to be incorporated

File: gpMgmt/bin/gppylib/programs/clsRecoverSegment.py [hunk 5]
```diff
@@ -370,7 +371,6 @@ def run(self):
 
         # figure out what needs to be done
         mirrorBuilder = self.getRecoveryActionsBasedOnOptions(gpEnv, gpArray)
-
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:373]: [nit]: Line is deleted

File: gpMgmt/bin/gppylib/utils.py [hunk 6]
```diff
@@ -640,3 +640,30 @@ def validateHostnameAddress(hostname, address):
...
+        if parts[0] in {"I", "D", "F", "i", "d", "f"}:
+            sync_mode = parts[0]
+            if sync_mode in {"D", "d"}:
+                recovery_type = "Differential"
+            elif sync_mode in {"F", "f"}:
+                recovery_type = "Full"
+        else:
+            hostname = parts[0]
+            hostname_check_required = True
+    if len(parts) == 5:
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/utils.py:665]: In the same function we already have a check for 'len == 5', can we re-organize code as below with comments to make it easier to understand.

```
recovery_type_flag = False
//comments
if len(parts) == 5:
     if parts[0] not in {"I", "D", "F", "i", "d", "f"}:
        raise Exception("Invalid recovery type provided, please provide any of I,D,F,i,d,f as recovery_type")
     hostname = parts[1]
     hostname_check_required = True
     recovery_type_flag = True

//comments
 if len(parts) == 4:
     if parts[0] in {"I", "D", "F", "i", "d", "f"}:
        recovery_type_flag = True
     else:
        hostname = parts[0]
        hostname_check_required = True

if recovery_type_flag:
    sync_mode = parts[0]
    if sync_mode in {"D", "d"}:
        recovery_type = "Differential"
    elif sync_mode in {"F", "f"}:
        recovery_type = "Full"
```

File: gpMgmt/bin/gppylib/programs/clsRecoverSegment.py [hunk 7]
```diff
@@ -405,7 +405,6 @@ def run(self):
             #TODO this already happens in buildMirrors function
             mirrorBuilder.checkForPortAndDirectoryConflicts(gpArray)
             self.validate_heap_checksum_consistency(gpArray, mirrorBuilder)
-
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:408]: [nit]: Line deleted

File: gpMgmt/bin/gppylib/operations/buildMirrorSegments.py [hunk 8]
```diff
@@ -110,16 +110,23 @@ def __init__(self, failedSegment, liveSegment, failoverSegment, forceFullSynchro
...
         the delta between the source and target datadir will be copied over to the target server
         """
-
+        self.__forceFullSynchronization = forceFullSynchronization
         self.__differentialSynchronization = differentialSynchronization
 
+        if not (
+            forceFullSynchronization or differentialSynchronization) and recoveryType in [
+            "Differential", "Full"] and self.__failoverSegment is None:
+            # If either forceFullSynchronization or differentialSynchronization is explicitly not set, and
```
Comment (ShirishaRao) [inline gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:123]: [nit]: Can we move the whole not check in one line to increase readability ?
```
if not (forceFullSynchronization or differentialSynchronization) and \
            recoveryType in ["Differential", "Full"] and self.__failoverSegment is None:
```

File: gpMgmt/bin/gppylib/utils.py (hunk 3 above)
Comment (Annu149) [reply gpMgmt/bin/gppylib/utils.py]: It is rare that hostname will be I/D/F/i/d/f but we will update the docs to make sure the user knows this one conflicting scenario and uses the hostname accordingly

**Review by ShirishaRao — APPROVED** [review]
Tested multiple scenarios, working as expected.

File: gpMgmt/bin/gppylib/utils.py [hunk 9]
```diff
@@ -640,3 +640,40 @@ def validateHostnameAddress(hostname, address):
                 address, hostname, resolved_address_list))
         return False
     return True
+
+def extract_recovery_config_info(parts):
+    """
+    Extracts relevant recovery configuration information from a list of parts.
+    """
+    address, port, datadir = parts[-3:]
```
Comment (shrakesh) [inline gpMgmt/bin/gppylib/utils.py:648]: as now this is standalone function, IMO sanity check or validation of input is also required here. 
if some how empty parts pass in the function the behaviour would be unexpected when accessing element if the list.

File: gpMgmt/bin/gppylib/utils.py [hunk 10]
```diff
@@ -640,3 +640,40 @@ def validateHostnameAddress(hostname, address):
...
+    recovery_type = "Incremental"
+    hostname_check_required = False
+    hostname = address
+
+    recovery_type_flag = False
+    # If 5 fields are there in part then first field represents recovery type and it should any of I,D,F.i,d,f
+    if len(parts) == 5:
```
Comment (shrakesh) [inline gpMgmt/bin/gppylib/utils.py:655]: in the caller section if the length of the parts is in 3,4,5 we are treating that as a valid input, in the function we are addressing len(parts) == 5 and len(parts) ==4. 
is that len(parts) ==3 is the default section ??

Comment (Annu149) [reply gpMgmt/bin/gppylib/utils.py:655]: yes

**Review by shrakesh — APPROVED** [review]
LGTM

**Review by nimish350 — APPROVED** [review]
Tested with i/f/d and I/F/D. LGTM+1

**Pull request merged**